File: app/src/main/java/com/verumomnis/forensic/engine/FindingsJsonEmitter.kt
package com.verumomnis.forensic.engine

import com.verumomnis.forensic.model.Contradiction
import com.verumomnis.forensic.model.ForensicFindings
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import java.time.Instant

/**
 * Engine-to-G3 contract serializer for the G3 Hybrid Report Pipeline (GHRP).
 *
 * Serializes deterministic [ForensicFindings] into the Findings JSON schema
 * (v1.0.0) that Gemma 3 narrates from. Every record is anchored to a source
 * document, page/line, and SHA-512 hash. G3-raised candidates are recorded in
 * the identical format but labelled as pending verification.
 */
object FindingsJsonEmitter {

    const val FINDINGS_JSON_VERSION = "1.0.0"
    const val ENGINE_VERSION = com.verumomnis.forensic.engine.contradiction.EngineVersion.VALUE

    const val STATUS_ENGINE_VERIFIED = "ENGINE-VERIFIED"
    const val STATUS_G3_CANDIDATE = "G3-RAISED CANDIDATE - PENDING VERIFICATION"
    const val STATUS_CANDIDATE_PROMOTED = "CANDIDATE PROMOTED - ENGINE-VERIFIED"
    const val STATUS_CANDIDATE_REJECTED = "CANDIDATE REJECTED - REASON LOGGED"

    @Serializable
    data class FindingsJsonDocument(
        val engine_version: String,
        val findings_json_version: String,
        val generated_utc: String,
        val source_bundle: String,
        val case_id: String = "",
        val case_ids: List<String> = emptyList(),
        val corpus_sha512: String = "",
        val engine_verified_count: Int = 0,
        val g3_candidate_count: Int = 0,
        val supplement_date: String? = null,
        val new_contradictions: Int? = null,
        val re_anchored_known: Int? = null,
        val integrity_findings: List<String> = emptyList(),
        val contradictions: List<FindingsJsonRecord>
    )

    @Serializable
    data class FindingsJsonRecord(
        val contradiction_id: String,
        val type: String,
        val severity: String,
        val confidence: String,
        val proposition_a_text: String,
        val proposition_a_actor: String,
        val proposition_b_text: String,
        val proposition_b_actor: String,
        val conflict_description: String,
        val source_document: String = "",
        val source_page: Int = 0,
        val source_line: Int = 0,
        val sha512_anchor: String = "",
        val extraction_method: String = "",
        val temporal_analysis: JsonElement? = null,
        val detected_fact: JsonElement? = null,
        val logical_pattern: JsonElement? = null,
        val legal_hypothesis: JsonElement? = null,
        val verification_status: String
    )

    /** Vault filename for the findings JSON artefact produced by a scan. */
    fun findingsFileName(caseName: String, now: Instant): String {
        val timestamp = now.toString().take(19).replace(":", "-")
        return "findings_${caseName}_${timestamp}.json"
    }

    /**
     * Serialize a full [ForensicFindings] object into the findings JSON contract.
     *
     * [extraRecords] carries G3-raised candidates (already converted to this
     * schema via [fromContractRecord]); the tier counts are recomputed so the
     * document always states how many findings are engine-verified versus
     * pending candidates.
     */
    fun emit(
        findings: ForensicFindings,
        caseName: String,
        now: Instant,
        corpusSha512: String = "",
        extraRecords: List<FindingsJsonRecord> = emptyList()
    ): String {
        val records = findings.contradictions.map { contradictionToRecord(it) } + extraRecords
        val candidateCount = records.count { it.verification_status == STATUS_G3_CANDIDATE }
        val document = FindingsJsonDocument(
            engine_version = ENGINE_VERSION,
            findings_json_version = FINDINGS_JSON_VERSION,
            generated_utc = now.toString(),
            source_bundle = caseName,
            case_id = caseName,
            case_ids = listOf(caseName),
            corpus_sha512 = corpusSha512,
            engine_verified_count = records.size - candidateCount,
            g3_candidate_count = candidateCount,
            integrity_findings = emptyList(),
            contradictions = records
        )
        return Json.encodeToString(document)
    }

    /**
     * Bridge from the GHRP contract record type produced by
     * [com.verumomnis.forensic.engine.contradiction.G3CandidateRegistry] into
     * this emitter's record type, so registry candidates merge into the
     * production findings document. One schema in the vault, two producers.
     */
    fun fromContractRecord(
        record: com.verumomnis.forensic.engine.contradiction.FindingsJsonEmitter.FindingsContradictionRecord
    ): FindingsJsonRecord = FindingsJsonRecord(
        contradiction_id = record.contradictionId,
        type = record.type,
        severity = record.severity,
        confidence = record.confidence,
        proposition_a_text = record.propositionAText,
        proposition_a_actor = record.propositionAActor,
        proposition_b_text = record.propositionBText,
        proposition_b_actor = record.propositionBActor,
        conflict_description = record.conflictDescription,
        source_document = record.sourceDocument,
        source_page = record.sourcePage,
        source_line = record.sourceLine,
        sha512_anchor = record.sha512Anchor,
        extraction_method = record.extractionMethod,
        temporal_analysis = null,
        detected_fact = null,
        logical_pattern = null,
        legal_hypothesis = null,
        verification_status = record.verificationStatus
    )

    /** Convert one deterministic-engine [Contradiction] into a findings JSON record. */
    fun contradictionToRecord(c: Contradiction): FindingsJsonRecord {
        val type = c.engineType.takeIf { it.isNotBlank() } ?: c.type.name
        val actorA = c.respondent.takeIf { it.isNotBlank() } ?: c.claimA.source
        val actorB = c.respondent.takeIf { it.isNotBlank() } ?: c.claimB.source

        val detectedFact = buildJsonObject {
            put("fact_text", JsonPrimitive(c.description.ifBlank { c.claimA.text }))
            put("source_document", JsonPrimitive(c.claimA.source))
            put("source_page", JsonPrimitive(c.claimA.page))
            put("source_line", JsonPrimitive(c.claimA.line))
            put("sha512_hash", JsonPrimitive(c.claimA.sha512))
            put("extraction_method", JsonPrimitive("engine"))
            put("confidence", JsonPrimitive(c.confidence.name))
        }

        val logicalPattern = buildJsonObject {
            put("pattern_type", JsonPrimitive(type))
            put("pattern_description", JsonPrimitive(c.description.ifBlank { "Contradiction between ${c.claimA.source} and ${c.claimB.source}" }))
            put("supporting_facts", buildJsonArray {
                add(JsonPrimitive(c.claimA.text))
                add(JsonPrimitive(c.claimB.text))
            })
            put("contradiction_score", JsonNull)
            put("detector_version", JsonPrimitive(ENGINE_VERSION))
        }

        return FindingsJsonRecord(
            contradiction_id = c.contradictionId,
            type = type,
            severity = c.severity.name,
            confidence = c.confidence.name,
            proposition_a_text = c.claimA.text,
            proposition_a_actor = actorA,
            proposition_b_text = c.claimB.text,
            proposition_b_actor = actorB,
            conflict_description = c.description,
            source_document = c.claimA.source,
            source_page = c.claimA.page,
            source_line = c.claimA.line,
            sha512_anchor = c.claimA.sha512,
            extraction_method = "engine",
            temporal_analysis = null,
            detected_fact = detectedFact,
            logical_pattern = logicalPattern,
            legal_hypothesis = null,
            verification_status = STATUS_ENGINE_VERIFIED
        )
    }

    /**
     * Record a contradiction G3 spotted in the sealed vault that the engine missed.
     *
     * The returned record is labelled [STATUS_G3_CANDIDATE] so it can never be
     * mistaken for an engine-verified finding. Promotion to engine-verified status
     * happens only after engine re-run or human sign-off.
     */
    fun raiseG3Candidate(
        candidateId: String,
        contradictionType: String,
        propositionAText: String,
        propositionBText: String,
        propositionAActor: String,
        propositionBActor: String,
        conflictDescription: String,
        sourceDocument: String,
        sourcePage: Int,
        sha512Anchor: String,
        severity: String = "MODERATE",
        confidence: String = "MODERATE",
        sourceLine: Int = 0,
        g3Model: String = "gemma-3-4b-it"
    ): FindingsJsonRecord {
        val extractionMethod = "G3 vault review ($g3Model) over sealed bundle"
        val detectedFact = buildJsonObject {
            put("fact_text", JsonPrimitive(conflictDescription))
            put("source_document", JsonPrimitive(sourceDocument))
            put("source_page", JsonPrimitive(sourcePage))
            put("source_line", JsonPrimitive(sourceLine))
            put("sha512_hash", JsonPrimitive(sha512Anchor))
            put("extraction_method", JsonPrimitive(extractionMethod))
            put("confidence", JsonPrimitive(confidence))
        }
        val logicalPattern = buildJsonObject {
            put("pattern_type", JsonPrimitive(contradictionType))
            put("pattern_description", JsonPrimitive(conflictDescription))
            put("supporting_facts", buildJsonArray {
                add(JsonPrimitive(propositionAText))
                add(JsonPrimitive(propositionBText))
            })
            put("contradiction_score", JsonNull)
            put("detector_version", JsonPrimitive("G3-CANDIDATE ($g3Model)"))
        }
        return FindingsJsonRecord(
            contradiction_id = candidateId,
            type = contradictionType,
            severity = severity,
            confidence = confidence,
            proposition_a_text = propositionAText,
            proposition_a_actor = propositionAActor,
            proposition_b_text = propositionBText,
            proposition_b_actor = propositionBActor,
            conflict_description = conflictDescription,
            source_document = sourceDocument,
            source_page = sourcePage,
            source_line = sourceLine,
            sha512_anchor = sha512Anchor,
            extraction_method = extractionMethod,
            temporal_analysis = null,
            detected_fact = detectedFact,
            logical_pattern = logicalPattern,
            legal_hypothesis = null,
            verification_status = STATUS_G3_CANDIDATE
        )
    }
}

File: findings_json_emitter.py
#!/usr/bin/env python3
"""
Verum Omnis - Findings JSON Emitter v1.0.0
===========================================
Engine-to-G3 contract serializer for the G3 Hybrid Report Pipeline (GHRP).

Status: PROPOSED - pending founder ratification (see G3_HYBRID_REPORT_PIPELINE.md)

What this does:
- Serializes engine Contradiction objects (verum_contradiction_engine_v531c.py)
  into the Findings JSON contract that Gemma 3 narrates from.
- Provides raise_g3_candidate() so contradictions G3 spots in the sealed vault
  are recorded in the identical format, labelled as candidates pending verification.

What this does NOT do:
- It does not modify detection logic. B1-B8 remain the deterministic spine.
- It does not hash artefacts. Hashing stays in the engine/vault (determinism);
  anchors are passed in as pre-computed SHA-512 strings.
- It never touches raw documents. Input is the engine's sealed output only.
"""

from __future__ import annotations

import json
from dataclasses import asdict, is_dataclass
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Dict, Iterable, List, Optional

FINDINGS_JSON_VERSION = "1.0.0"

STATUS_ENGINE_VERIFIED = "ENGINE-VERIFIED"
STATUS_G3_CANDIDATE = "G3-RAISED CANDIDATE - PENDING VERIFICATION"
STATUS_CANDIDATE_PROMOTED = "CANDIDATE PROMOTED - ENGINE-VERIFIED"
STATUS_CANDIDATE_REJECTED = "CANDIDATE REJECTED - REASON LOGGED"


def _enum_value(v: Any) -> Any:
    """Unwrap Enum members to their .value for clean JSON."""
    if isinstance(v, Enum):
        return v.value
    return v


def _obj_to_dict(obj: Any) -> Any:
    """Dataclass / dict / Enum-safe conversion."""
    if obj is None:
        return None
    if isinstance(obj, Enum):
        return obj.value
    if is_dataclass(obj) and not isinstance(obj, type):
        return {k: _obj_to_dict(v) for k, v in asdict(obj).items()}
    if isinstance(obj, dict):
        return {k: _obj_to_dict(v) for k, v in obj.items()}
    if isinstance(obj, (list, tuple)):
        return [_obj_to_dict(v) for v in obj]
    return obj


def contradiction_to_record(c: Any) -> Dict[str, Any]:
    """
    Convert one engine Contradiction (dataclass instance or plain dict) into a
    Findings JSON record. Tolerant of both shapes so sealed/legacy dicts also emit.
    """
    d = _obj_to_dict(c) if not isinstance(c, dict) else dict(c)

    df = d.get("detected_fact") or {}
    lp = d.get("logical_pattern") or {}

    record = {
        "contradiction_id": d.get("contradiction_id", ""),
        "type": _enum_value(d.get("type", "")),
        "severity": _enum_value(d.get("severity", "")),
        "confidence": _enum_value(d.get("confidence", "")),
        "proposition_a_text": d.get("proposition_a_text", ""),
        "proposition_a_actor": d.get("proposition_a_actor", ""),
        "proposition_b_text": d.get("proposition_b_text", ""),
        "proposition_b_actor": d.get("proposition_b_actor", ""),
        "conflict_description": d.get("conflict_description", ""),
        "source_document": df.get("source_document", ""),
        "source_page": df.get("source_page", 0),
        "source_line": df.get("source_line", 0),
        "sha512_anchor": df.get("sha512_hash", ""),
        "extraction_method": df.get("extraction_method", ""),
        "temporal_analysis": d.get("temporal_analysis"),
        "detected_fact": df or None,
        "logical_pattern": lp or None,
        "legal_hypothesis": d.get("legal_hypothesis"),
        "verification_status": d.get("verification_status") or STATUS_ENGINE_VERIFIED,
    }
    return record


def raise_g3_candidate(
    *,
    candidate_id: str,
    contradiction_type: str,
    proposition_a_text: str,
    proposition_b_text: str,
    proposition_a_actor: str,
    proposition_b_actor: str,
    conflict_description: str,
    source_document: str,
    source_page: int,
    sha512_anchor: str,
    severity: str = "MODERATE",
    confidence: str = "MODERATE",
    source_line: int = 0,
    g3_model: str = "gemma-3-4b-it",
) -> Dict[str, Any]:
    """
    Record a contradiction G3 spotted in the sealed vault that the engine missed.

    Two-tier rule (GHRP section 4): candidates are anchored and hashed like any
    other artefact, but labelled pending verification. They never masquerade as
    engine-verified. Promotion happens via engine re-run or human sign-off, and
    is recorded by flipping verification_status to STATUS_CANDIDATE_PROMOTED.
    """
    return {
        "contradiction_id": candidate_id,
        "type": contradiction_type,
        "severity": severity,
        "confidence": confidence,
        "proposition_a_text": proposition_a_text,
        "proposition_a_actor": proposition_a_actor,
        "proposition_b_text": proposition_b_text,
        "proposition_b_actor": proposition_b_actor,
        "conflict_description": conflict_description,
        "source_document": source_document,
        "source_page": source_page,
        "source_line": source_line,
        "sha512_anchor": sha512_anchor,
        "extraction_method": f"G3 vault review ({g3_model}) over sealed bundle",
        "temporal_analysis": None,
        "detected_fact": {
            "fact_text": conflict_description,
            "source_document": source_document,
            "source_page": source_page,
            "source_line": source_line,
            "sha512_hash": sha512_anchor,
            "extraction_method": f"G3 vault review ({g3_model}) over sealed bundle",
            "confidence": confidence,
        },
        "logical_pattern": {
            "pattern_type": contradiction_type,
            "pattern_description": conflict_description,
            "supporting_facts": [proposition_a_text, proposition_b_text],
            "contradiction_score": None,
            "detector_version": f"G3-CANDIDATE ({g3_model})",
        },
        "legal_hypothesis": None,
        "verification_status": STATUS_G3_CANDIDATE,
    }


def emit_findings_json(
    contradictions: Iterable[Any],
    *,
    engine_version: str,
    source_bundle: str,
    case_ids: Optional[List[str]] = None,
    supplement_date: Optional[str] = None,
    integrity_findings: Optional[List[str]] = None,
    extra_records: Optional[Iterable[Dict[str, Any]]] = None,
) -> Dict[str, Any]:
    """
    Build the complete Findings JSON document for one scan.

    `extra_records` accepts G3 candidates produced by raise_g3_candidate().
    Counts split engine-verified vs candidate tiers for the report header.
    """
    records: List[Dict[str, Any]] = [contradiction_to_record(c) for c in contradictions]
    if extra_records:
        records.extend(extra_records)

    candidate_count = sum(
        1 for r in records if r.get("verification_status") == STATUS_G3_CANDIDATE
    )

    return {
        "engine_version": engine_version,
        "findings_json_version": FINDINGS_JSON_VERSION,
        "generated_utc": datetime.now(timezone.utc).isoformat(),
        "source_bundle": source_bundle,
        "case_ids": case_ids or [],
        "supplement_date": supplement_date,
        "engine_verified_count": len(records) - candidate_count,
        "g3_candidate_count": candidate_count,
        "integrity_findings": integrity_findings or [],
        "contradictions": records,
    }


def write_findings_json(findings: Dict[str, Any], output_path: str) -> str:
    """Write Findings JSON to disk (UTF-8, stable formatting). Returns the path."""
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(findings, f, indent=2, ensure_ascii=False)
    return output_path


if __name__ == "__main__":
    # Smoke test: one engine-style record + one G3 candidate.
    demo_engine_record = {
        "contradiction_id": "VO-DEMO-001",
        "type": "STATEMENT_VS_STATEMENT",
        "severity": "HIGH",
        "confidence": "HIGH",
        "proposition_a_text": "Statement A",
        "proposition_a_actor": "Actor A",
        "proposition_b_text": "Statement B",
        "proposition_b_actor": "Actor B",
        "conflict_description": "A and B cannot both be true.",
        "detected_fact": {
            "source_document": "demo.pdf",
            "source_page": 1,
            "source_line": 3,
            "sha512_hash": "ab" * 64,
            "extraction_method": "engine",
            "confidence": "HIGH",
        },
    }
    demo_candidate = raise_g3_candidate(
        candidate_id="VO-DEMO-G3-001",
        contradiction_type="OMISSION",
        proposition_a_text="Reply promised by next week",
        proposition_b_text="No reply exists in sealed record",
        proposition_a_actor="Institution",
        proposition_b_actor="Sealed record",
        conflict_description="Undertaking never honoured anywhere in the vault.",
        source_document="demo.pdf",
        source_page=7,
        sha512_anchor="cd" * 64,
    )
    out = emit_findings_json(
        [demo_engine_record],
        engine_version="5.3.1c",
        source_bundle="demo.pdf",
        case_ids=["demo"],
        extra_records=[demo_candidate],
    )
    print(json.dumps(out, indent=2))
